bitcoin/src/blockdata/constants.rs

```diff
@@ -275,12 +286,12 @@ mod test {
         use hashes::sha256;
 
         // The genesis block hash is a double-sha256 and it is displayed backwards.
-        let genesis_hash = genesis_block(network).block_hash();
+        let genesis_hash = genesis_block(&Params::from(network)).block_hash();
```

Comment:
In a1637ecdc793177a41b9ea83063d3aa0ab30d582:

Is this change required? Users shouldn't have to change their existing code (except maybe to stick an `&` on their network, though I hope even that isn't needed). They certainly shouldn't need to write `Params::from`.

Reply:
Yeah I also thought this change is ugly. A way around it is to implement `AsRef<Params> for Network` but I was, and still am, unsure if that is correct usage of `AsRef` and unsurprising?

Reply:
We should either implement `AsRef

The docs suggest we actually want `Into<&Params>` or maybe `Into<Params>` https://doc.rust-lang.org/std/convert/trait.AsRef.html for "nontrivial conversions".

But I don't think `AsRef<Params> for Network` would be *that* surprising or wrong, no.

Reply:
Yeah, my personal indecision is based on whether or not `network.as_ref()` is trivial or not? More reading just now showed me that I was mis-understanding `const` a bit. `const` variables get inlined so `network.as_ref()` is not really trivial because it leads to a bunch of additional data on the stack, right? Should we be using `static` for `Params` instead, then `network.as_ref()` is just a trivial reference to the single instance of `Params` (for said network).

Reply:
Yeah, we might as well, but I think global `const`s are the same as global `static`s. I can't imagine that rustc would be copying a global `const` onto the stack anywhere, and your link doesn't suggest that it would.

Reply:
huh? 

> More specifically, constants in Rust have no fixed address in memory. This is because they’re effectively inlined to each place that they’re used.

Reply:
Oh, apparently I just can't read :)

Reply:
I added 
```rust
/// The mainnet parameters.
///
/// Use this for a static reference eg, `&params::MAINNET`.
///
/// For more on static vs const see The Rust Reference [using-statics-or-consts] section.
///
/// [using-statics-or-consts]: <https://doc.rust-lang.org/reference/items/static-items.html#using-statics-or-consts>
pub static MAINNET: Params = Params::MAINNET;
/// The testnet parameters.
pub static TESTNET: Params = Params::TESTNET;
/// The signet parameters.
pub static SIGNET: Params = Params::SIGNET;
/// The regtest parameters.
pub static REGTEST: Params = Params::REGTEST;

...

impl AsRef<Params> for Network {
    fn as_ref(&self) -> &Params {
        match *self {
            Network::Bitcoin => &MAINNET,
            Network::Testnet => &TESTNET,
            Network::Signet => &SIGNET,
            Network::Regtest => &REGTEST,
        }
    }
}
```

Reply:
And used `&params::FOO` everywhere.

Reply:
We should leave `Network::FOO` in a couple places just to make sure it keeps working.

Reply:
Ah yes! Makes the diff smaller also, why didn't I think of that.